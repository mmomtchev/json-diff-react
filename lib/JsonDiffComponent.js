var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { diffRender } from './JsonDiff/Internal/index';
var defaultStyleCustomization = {
    additionLineStyle: {
        color: 'green',
        lineHeight: 0.5,
    },
    additionClassName: 'addition',
    deletionLineStyle: {
        color: 'red',
        lineHeight: 0.5,
    },
    deletionClassName: 'deletion',
    unchangedLineStyle: {
        lineHeight: 0.5,
    },
    unchangedClassName: 'unchanged',
    frameStyle: {},
    frameClassName: 'diff',
};
export function mkCustomization(customizations) {
    return __assign(__assign({}, defaultStyleCustomization), customizations);
}
/**
 * React.js functional component for a structural JSON diff.
 *
 * @param {Object} props - properties of the component
 * @param {JsonValue} props.original - parsed JSON value
 * @param {JsonValue} props.latest - parsed JSON value
 * @param {Partial<StyleCustomization>=} props.styleCustomization
 * @param {DiffOptions} props.jsonDiffOptions - properties passed to json-diff
 */
export function JsonDiffComponent(props) {
    var _a;
    var actualCustomization = (_a = props.styleCustomization) !== null && _a !== void 0 ? _a : {};
    var fullCustomization = mkCustomization(actualCustomization);
    var diffElement = diffRender(props.original, props.latest, props.jsonDiffOptions, fullCustomization);
    return _jsx("div", { children: diffElement });
}
