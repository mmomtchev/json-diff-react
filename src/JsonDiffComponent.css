.deletion {
  /* customization for deleted lines (div) */
}

.addition {
  /* customization for added lines (div) */
}

.unchanged {
  /* customization for unchanged lines (div) */
}

.diff {
  /* customization for the top-level <div> wrapper that contains the diff */
}
