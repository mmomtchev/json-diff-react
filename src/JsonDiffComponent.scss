.deletion {
  /* customization for deleted lines */
}

.addition {
  /* customization for added lines */
}

.unchanged {
  /* customization for unchanged lines */
}
