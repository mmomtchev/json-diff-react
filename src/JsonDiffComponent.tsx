import { diffRender, JsonValue, DiffOptions, StyleCustomization } from './JsonDiff/Internal/index';

// Default style customization only sets class names to their default values.
const defaultStyleCustomization: StyleCustomization = {
  additionLineStyle: null,
  additionClassName: 'addition',
  deletionLineStyle: null,
  deletionClassName: 'deletion',
  unchangedLineStyle: null,
  unchangedClassName: 'unchanged',
  frameStyle: null,
  frameClassName: 'diff',
};

export function mkCustomization(customizations: Partial<StyleCustomization>): StyleCustomization {
  return { ...defaultStyleCustomization, ...customizations };
}

/**
 * React.js functional component for a structural JSON diff.
 *
 * @param {Object} props - properties of the component
 * @param {JsonValue} props.jsonA - parsed JSON value (fed to diff)
 * @param {JsonValue} props.jsonB - parsed JSON value (fed to diff)
 * @param {Partial<StyleCustomization>} props.styleCustomization
 * @param {DiffOptions} props.jsonDiffOptions - properties passed to json-diff
 */
export function JsonDiffComponent(props: {
  jsonA: JsonValue;
  jsonB: JsonValue;
  styleCustomization?: Partial<StyleCustomization>;
  jsonDiffOptions?: DiffOptions;
}): JSX.Element {
  return diffRender(
    props.jsonA,
    props.jsonB,
    props.jsonDiffOptions,
    mkCustomization(props.styleCustomization ?? {})
  );
}

// Re-export 'StyleCustomization' and 'DiffOptions'
export type { StyleCustomization, DiffOptions, JsonValue } from './JsonDiff/Internal/index';
