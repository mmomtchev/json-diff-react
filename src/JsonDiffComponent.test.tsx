import { describe, expect } from '@jest/globals';
import * as fc from 'fast-check';
import renderer from 'react-test-renderer';

import React from 'react';

import { JsonDiffComponent, StyleCustomization } from './JsonDiffComponent';

const runSnapshotTest = (
  original: string, 
  latest: string, 
  onError: (e: Error) => JSX.Element, 
  styleCustomization: Partial<StyleCustomization>
  ) => {

  const tree = renderer
    .create(
      <JsonDiffComponent 
        original={original} 
        latest={latest} 
        onError={onError} 
        styleCustomization={styleCustomization}
        />)
    .toJSON();

  expect(tree).toMatchSnapshot();
  expect(onError).not.toHaveBeenCalled();
}; 

const runSimpleSnapshotTest = (original: string, latest: string) => {
  runSnapshotTest(original, latest, jest.fn(), {});
};

describe('<JsonDiffComponent />', () => {
  it('it should render correctly (objects vs object)', () => {
    const original = {
      key1: "value1",
      key2: 123,
      key3: [1, 2, 3]
    };

    const latest = {
      key1: "value",
      key2: 123,
      key3: [1, 2, 3, 4]
    };
    
    runSimpleSnapshotTest(JSON.stringify(original), JSON.stringify(latest));
  });

  it('it should render correctly (array vs array)', () => {
    const original = [1, 2, 3];
    const latest = [1, 2, 3, 4];
    
    runSimpleSnapshotTest(JSON.stringify(original), JSON.stringify(latest));
  });

  it('it should render correctly (object vs array)', () => {
    const original = {
      key1: "value1",
      key2: 123,
      key3: [1, 2, 3]
    };
    const latest = [1, 2, 3, 4];
    
    runSimpleSnapshotTest(JSON.stringify(original), JSON.stringify(latest));
  });

  it('should pick up CSS classes from styleCustomization', () => {
    const original = { key1: "value1", key2: 123, key3: [1, 2, 3] };
    const latest = { key1: "value", key2: 123, key3: [1, 2, 3, 4] };

    const styleCustomization: Partial<StyleCustomization> = {
      additionClassName: "custom_addition_class",
      deletionClassName: "custom_deletion_class",
      unchangedClassName: "custom_unchanged_class",
      frameClassName: "custom_frame_class"
    }

    runSnapshotTest(
      JSON.stringify(original),
      JSON.stringify(latest),
      jest.fn(),
      styleCustomization 
    );
  });

  it('should pick up CSS styles from styleCustomization', () => {
    const original = { key1: "value1", key2: 123, key3: [1, 2, 3] };
    const latest = { key1: "value", key2: 123, key3: [1, 2, 3, 4] };

    const styleCustomization: Partial<StyleCustomization> = {
      additionLineStyle: { lineHeight: 1 },
      deletionLineStyle: { lineHeight: 2 },
      unchangedLineStyle: { lineHeight: 3 },
      frameStyle: { backgroundColor: "black" }
    }

    runSnapshotTest(
      JSON.stringify(original),
      JSON.stringify(latest),
      jest.fn(),
      styleCustomization 
    );
  });

  it('should return an empty div if there are no changes (snapshot)', () => {
    runSimpleSnapshotTest("{}", "{}");
  });

  it('should return an empty div if there are no changes (property)', () => {
    fc.assert(
      fc.property(fc.json(), (json) => {
        const onError = jest.fn();
        const tree = renderer
          .create(<JsonDiffComponent original={json} latest={json} onError={onError} />)
          .toJSON();

        expect(onError).not.toHaveBeenCalled();
        expect(tree).toMatchInlineSnapshot(`
          <div>
            <div
              className="diff"
              style={{}}
            >
              <pre />
            </div>
          </div>
        `);
      })
    );
  });

  it('onError should not be called if both inputs are JSON encoded', () => {
    fc.assert(
      fc.property(fc.json(), fc.json(), (original, latest) => {
        const onError = jest.fn();
        renderer
          .create(<JsonDiffComponent original={original} latest={latest} onError={onError} />)
          .toJSON();

        expect(onError).not.toHaveBeenCalled();
      })
    );
  });

  it('onError should be called if either input is not valid JSON', () => {
    function isValidJSON(str: string): boolean {
      try {
        JSON.parse(str);
      } catch(e) {
        return false;
      }

      return true;
    }

    fc.assert(
      fc.property(fc.string(), fc.string(), (original, latest) => {
        fc.pre(!isValidJSON(original) || isValidJSON(latest)); 

        const onError = jest.fn();
        renderer
          .create(<JsonDiffComponent original={original} latest={latest} onError={onError} />)
          .toJSON();

        expect(onError).toHaveBeenCalled();
      })
    );
  });

});

