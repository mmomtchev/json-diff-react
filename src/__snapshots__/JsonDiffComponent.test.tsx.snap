// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JsonDiffComponent /> should return an empty div if there are no changes 1`] = `
<div>
  <div
    className="diff"
    style={{}}
  >
    <pre />
  </div>
</div>
`;
